### PR TITLE
Use protobuffer instead of JSON

### PR DESCRIPTION
**Pull request by AgustinCB**
The objective of this commit it's to use protobuffer instead of JSON.
This PR doesn't remove JSON completely from the project, there are three
corners in which it's still present:

- src/service : It's present here because this is a public facing HTTP
service, in the form of a REST API. Using JSON here makes sense.

- src/peers/json_peers.go : This is intended as a way to serialize in
disk a datastructure in a format that can be read and modified by
humans. Again, using JSON here makes sense.

- src/net : Ideally, this would use protobuffers too. However, the way
the code is structured makes it hard. Encoders and decoders are used for
`interface{}` typed variables with no kind of schema. To use protobuffer
there we would have to first change the code to use strongly typed
values and I don't have enough context to know what's the best structure
to use there. An issue covering this should be opened.

Some comments:

We use a lot of `int`, which are an alias for "either int32 or int64
depending on architecture". Protobuf doesn't have that concept (of
course!), so I have to disambiguate that. I decided to go with int64
because I'm assuming that most tests were made in 64 bits architecture
so far and there it's less likely to unintentionally break things.

I had to start dealing with a bunch of conversions to and from `int` and
`int64`. This wasn't trivial, as `int` is used everywhere in code and
`int64` is used everywhere in protobuf. I'm doing the bare minimum to
make this work, which includes accepting that there'll still be some
`int`'s in the repo and that there will be some dangerous conversions
(i.e. a conversion from `int64` to `int`, which can overflow in a 32-bit
architecture). I created [this issue](https://github.com/andrecronje/lachesis/issues/60) to handle the
rest.

I created the `*.pb.go` files by using the following command: `protoc
--proto_path=src/poset/ --go_out=paths=source_relative:./src/poset
src/poset/*.proto`.

I had to modify the file event.pb.go. I know it's a bad idea, but it
seemed the only way to keep the code working. In particular, I had to
lower case the four last properties names in the type `Event`. Is there
a better way to do so? Worth creating a ticket documenting the technical
debt?

I had to change the functions that were using an alias (`[]Index`) as
the receiver object of a function and instead pass it as an explicit
parameter. That's because protobuffer changed the type to `[]*Index`
and that messed up goc, which wasn't able to find the functions.

Because we do a bunch of encoding of `map[string]int` and there's no way
to do just that with protobuf, I added a new structure, a simple wrapper
around that type, to allow it.

We had a bunch of pointers in place of integers. I'm guessing this is
because those attributes where optional (that way, `null` means not
set). However, such a concept doesn't exist in protobuffer. For now,
I'll assume that `int64` it's just fine and then see what breaks for not
making it optional. Another alternative would be to use `proto2` instead
of `proto3`, which allows the `required` tag, but which doesn't have the
type `map`. I decided not to go with that because using pointers to
integers seems hacky anyway.

We had a bunch of pointers to integers instead of full integers. That's
because we use the value `null` to express "not yet known." Protobuffer
doesn't support that kind of hack. In a language with `Option` this
would have a very easy solution, but given that Go handles nullable
values in a horrible way, that's out of our reach.

There's another problem with our integers: They're `int`, a type not
recognized by Protobuffer.

This PR converts those `*int` and `int` to `int64` and uses `-1` as "not
yet known." In a later discussion with Samuel, he mentioned that
there's a better solution: All those numbers (ids, rounds, timestamps)
can't be negative and should have been unsigned in the first place. We
should probably use `uin64` and use `!(0: uint64)` as "not yet known."
Although I agree with him, this PR is already getting pretty big and
making that change would make even bigger. In order to make this in
smaller, testable steps, I'm going to tackle that concern as part of
issue andrecronje/lachesis#60.

TL;DR:

In this PR:

- Change everything in poset to use protobuffer.
- Change all types in poset out of `int` into concrete size integers.

Pending to do:

- Update documentation with instructions on how to set up protobuffer on
a fresh repo.
- Convert every other `int` to `uint64` and `int64` to `uint64` (i.e.
issue #60).
- Refactor `src/net` to not use `interface{}` and use instead
protobuffer messages with defined schemas.

### CODE REVIEW

**Comment by AgustinCB** [comment]
Apologies for the huge PR, I promise the next will be smaller. I'd appreciate a detailed review of my changes, since they're very sensitive and I have very little knowledge of the project so far.

**Comment by SamuelMarks** [comment]
Thanks for your efforts.

@AgustinCB I merged with my master locally, ran this command:
```
TARGET_OS=darwin n=3 logshold=-1,-1,-1,5,0,0 entry=main BUILD_DIR="$PWD" ./scripts/multi.bash
```

Got:
```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow
```

[Full log](https://github.com/SamuelMarks/lachesis/files/2536942/stackoverflow.log)

**Comment by AgustinCB** [comment]
Uhm, didn't happen on my box. I have a mac, will reproduce on that tomorrow.

**Comment by SamuelMarks** [comment]
@AgustinCB that didn't work, and there are merge conflicts also.

[Full output](https://github.com/SamuelMarks/lachesis/files/2538554/log.log)

**Comment by AgustinCB** [comment]
Merge conflict fixed. Working on the new error.

**Comment by AgustinCB** [comment]
Found the error. Protobuffer adds some internal values onto structures the that are differently set when verifying.

Working on fix + test on that.

File: src/poset/store.go [hunk 1]
```diff
@@ -10,13 +10,13 @@ type Store interface {
...
-	ParticipantEvent(string, int) (string, error)
+	ParticipantEvents(string, int64) ([]string, error)
+	ParticipantEvent(string, int64) (string, error)
 	LastEventFrom(string) (string, bool, error)
 	LastConsensusEventFrom(string) (string, bool, error)
-	KnownEvents() map[int]int
+	KnownEvents() map[int64]int64
 	ConsensusEvents() []string
-	ConsensusEventsCount() int
+	ConsensusEventsCount() int64
```
Comment (SamuelMarks) [inline src/poset/store.go]: Why not `uint64`? - I see lots of `int64` in this commit, that should obviously be `unit64`…

Comment (AgustinCB) [reply src/poset/store.go]: I agree. Almost all the changes in this PR should be `s/int64/uint64/g`.

However, so far I'm trying to do the bare minimum to have working, ugly and all, and once I have that working I will refactor to make it more meaningful.

**Comment by AgustinCB** [comment]
After discussion with @SamuelMarks : It's better to start merging stuff to reduce conflicts.

After fighting with this till literally 3:15AM, I wasn't able to scratch from here a functional pr that doesn't break at least something. So I decided a different approach: I'll merge this is in smaller steps.

1. int -> int64.
2. Convert all `*int64` in `poset` to `int64`.
3. Create Protobuffer structures and use them with JSON.
4. JSON -> Protobuffer.

I'll leave the transition int64 -> uint64 as part of issue #60 upstream. Mostly because that require logical changes that introduce unnecessary complexity here.
